File: mini-extractor-go/lib/token.go
package lib

import (
	"strings"

	"github.com/ethereum/go-ethereum/common"
)

// Token represents a minimal ERC20 token.
type Token struct {
	ChainID  int
	Decimals int
	Symbol   string
	Name     string
	Address  common.Address
}

// SortsBefore returns true if this token's address is lower (lexicographic) than other.
func (t *Token) SortsBefore(other *Token) bool {
	return strings.ToLower(t.Address.Hex()) < strings.ToLower(other.Address.Hex())
}

File: mini-extractor-ts/lib/token.ts
import { type Address, getAddress } from 'viem'

/**
 * Minimal Token class
 */
export class Token {
  public readonly chainId: number
  public readonly decimals: number
  public readonly symbol?: string | undefined
  public readonly name?: string | undefined
  public readonly address: Address
  public readonly tokenURI?: string | undefined

  public constructor({
    chainId,
    address,
    decimals,
    symbol,
    name,
    tokenURI,
  }: {
    chainId: number | string
    address: string
    decimals: number | string
    symbol?: string | undefined
    name?: string | undefined
    tokenURI?: string | undefined
  }) {
    this.chainId = Number(chainId)
    this.decimals = Number(decimals)
    this.symbol = symbol
    this.name = name
    this.tokenURI = tokenURI
    this.address = getAddress(address)
  }

  public sortsBefore(other: Token): boolean {
    return this.address.toLowerCase() < other.address.toLowerCase()
  }
}
